File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.py
import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
stack = [0]

for a in arr:
    if stack[-1] < a:
        stack.append(a)
    else:
        start = 0
        end = len(stack)
        while(start < end):
            mid = (start + end) // 2

            if stack[mid] < a:
                start = mid + 1
            else:
                end = mid
        stack[end] = a
print(len(stack) - 1)

File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.java
import java.util.*;
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		int[] arr = new int[n];

		List<Integer> list = new ArrayList<>();
		list.add(0);

		StringTokenizer st = new StringTokenizer(br.readLine());
		for (int i = 0; i < n; i++){
			int value = Integer.parseInt(st.nextToken());
			arr[i] = value;
			if (value > list.get(list.size() - 1))
				list.add(value);
			else{
				int l = 0;
				int r = list.size() - 1;
				while (l < r) {
					int mid = (l + r) / 2;
					if (list.get(mid) < value){
						l = mid + 1;
					}
					else {
						r = mid;
					}
				}
				list.set(r, value);
			}
		}
		System.out.println(list.size() - 1);
	}
}